### PR TITLE
libhwy: 0.15.0 -> 0.17.0

### PR DESCRIPTION
**Pull request by ethindp**
###### Description of changes

Change log for libhwy versions 0.16.0 and 0.17.0: https://github.com/google/highway/releases/tag/0.16.0 and https://github.com/google/highway/releases/tag/0.17.0

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by K900** [comment]
This should supposedly fix SIGILL on devices with no AVX512.

**Review by zhaofengli — APPROVED** [review]

File: pkgs/development/libraries/libhwy/default.nix [hunk 1]
```diff
@@ -2,25 +2,15 @@
...
     repo = "highway";
     rev = version;
-    sha256 = "sha256-v2HyyHtBydr7QiI83DW1yRv2kWjUOGxFT6mmdrN9XPo=";
+    hash = "sha256-CHjLD2qXmmggJXm15tdp+Nc/vyUIlrg4ZOvI4FA9lkc=";
   };
 
-  patches = [
-    # Remove on next release
-    # https://github.com/google/highway/issues/460
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libhwy/default.nix:17]: Please remove the unused fetchpatch from inputs.

**Comment by ethindp** [comment]
I've removed fetchpatch as requested.

**Review by zhaofengli — APPROVED** [review]

**Comment by risicle** [comment]
Cherry-picked back to master for testing, this causes `libjxl` to fail to build on macos 10.15:

```
In file included from /tmp/nix-build-libjxl-0.6.1.drv-0/source/lib/jxl/dec_transforms_testonly.cc:10:
In file included from /nix/store/j1bh6hhjvb6gwzvfx3kzsm1rvyfcaayb-libhwy-0.17.0/include/hwy/foreach_target.h:136:
In file included from /tmp/nix-build-libjxl-0.6.1.drv-0/source/lib/jxl/dec_transforms_testonly.cc:14:
In file included from /tmp/nix-build-libjxl-0.6.1.drv-0/source/lib/jxl/dec_transforms-inl.h:19:
In file included from /tmp/nix-build-libjxl-0.6.1.drv-0/source/lib/jxl/dct-inl.h:21:
/tmp/nix-build-libjxl-0.6.1.drv-0/source/lib/jxl/transpose-inl.h:148:23: error: no matching function for call to 'InterleaveUpper'
```

this makes me nervous to backport this bump even once we resolve the above.

**Comment by ethindp** [comment]
@risicle It doesn't build for me either... Do we have to update libjxl as well?

**Comment by risicle** [comment]
Yup, similar for nixos x86_64:

```
In file included from /build/source/lib/jxl/dct-inl.h:21,
                 from /build/source/lib/jxl/dec_transforms-inl.h:19,
                 from /build/source/lib/jxl/dec_transforms_testonly.cc:14:
/build/source/lib/jxl/transpose-inl.h:153:38: error: 'InterleaveUpper' was not declared in this scope
  153 |       const auto r3 = InterleaveUpper(q2, q3);
      |                       ~~~~~~~~~~~~~~~^~~~~~~~
```

**Comment by risicle** [comment]
If we want the SIGILL/AVX512 fix for stable, we should try and locate the patch.

**Comment by ethindp** [comment]
So, I tried reverting to 0.16.0 with no success. There also doesn't appear to be an update to libjxl, so we may wish to hold off merging this until libjxl is updated. I'll also submit an issue if it doesn't already exist about libjxl
failing to compile with libhwy > 0.15.0.

**Comment by ethindp** [comment]
@risicle So, I have identified a solution.

As it currently stands, we have libjxl build using our version of libhwy. However, libhwy is also vendored within the libjxl repositories. Thus, upgrading libjxl to 0.7rc (the upcoming release candidate) and disabling that vendoring temporarily resolves the problem. Alternatively, upgrading both libjxl and libhwy to their latest versions (in this case, libjxl 0.7rc and libhwy 1.0.0) also resolves the problem. I can do the update to libhwy and libjxl, assuming Git wants to be friendly. (I attempted to do this but accidentally branched off of another branch by accident so will need to redo it.) I will however need to rename this PR, or I can open another depending on your thoughts.

**Comment by risicle** [comment]
(I'm not in charge here, but..) how long do you think the 0.7 libjxl release will stay in rc? For unstable I wouldn't be totally against going for the rc, but if it's just a matter of waiting a week to get a proper release, waiting might save everyone some time in the long run.

As for PRs, whatever's convenient for you.

**Comment by ethindp** [comment]
@risicle I am not familiar enough with the libjxl development process to be able to estimate a release date, TBH. Perhaps @nh2 might be able to provide a better answer. I didn't see anything about release schedules in the repository either, though I might've missed something.

**Comment by risicle** [comment]
What is *our* main pressing factor? The AVX512 issue?

**Comment by nagy** [comment]
libjxl 0.7.0 has been released https://github.com/libjxl/libjxl/releases/tag/v0.7.0

and also libhwy 1.0.1 https://github.com/google/highway/releases/tag/1.0.1

**Comment by lopsided98** [comment]
I tried to update to libhwy 1.0.2 but it fails to build on armv7l:
```
FAILED: CMakeFiles/hwy_contrib.dir/hwy/contrib/sort/vqsort_128a.cc.o 
/nix/store/yrjxdsvrb426l4g9hqj0jfmwwwz08dsi-gcc-wrapper-11.3.0/bin/g++ -DHWY_STATIC_DEFINE -I/build/source -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -fmerge-all-constants -Wall -Wextra -Wconversion -Wsign-conversion -Wvla -Wnon-virtual-dtor -fmath-errno -fno-exceptions -MD -MT CMakeFiles/hwy_contrib.dir/hwy/contrib/sort/vqsort_128a.cc.o -MF CMakeFiles/hwy_contrib.dir/hwy/contrib/sort/vqsort_128a.cc.o.d -o CMakeFiles/hwy_contrib.dir/hwy/contrib/sort/vqsort_128a.cc.o -c /build/source/hwy/contrib/sort/vqsort_128a.cc
In file included from /build/source/hwy/ops/arm_neon-inl.h:33,
                 from /build/source/hwy/highway.h:358,
                 from /build/source/hwy/contrib/sort/shared-inl.h:103,
                 from /build/source/hwy/contrib/sort/traits128-inl.h:27,
                 from /build/source/hwy/contrib/sort/vqsort_128a.cc:23,
                 from /build/source/hwy/foreach_target.h:81,
                 from /build/source/hwy/contrib/sort/vqsort_128a.cc:20:
/nix/store/p3yqxgdqif9rws6fqsqr9bl2vi1rdnyw-gcc-11.3.0/lib/gcc/armv7l-unknown-linux-gnueabihf/11.3.0/include/arm_neon.h: In function 'void hwy::N_NEON::StoreU(hwy::N_NEON::Vec128<long long unsigned int, 2>, hwy::N_NEON::Full128<long long unsigned int>, uint64_t*)':
/nix/store/p3yqxgdqif9rws6fqsqr9bl2vi1rdnyw-gcc-11.3.0/lib/gcc/armv7l-unknown-linux-gnueabihf/11.3.0/include/arm_neon.h:11002:1: error: inlining failed in call to 'always_inline' 'void vst1q_u64(uint64_t*, uint64x2_t)': target specific option mismatch
```

This is an upstream bug, but there hasn't been a patch or workaround except to enable NEON support: https://github.com/google/highway/issues/834. libjxl looks to have a similar issue as well: https://github.com/libjxl/libjxl/issues/1748

**Pull request closed without merging**